### PR TITLE
refactor(kernels): split ffi.rs and reorganize csrc/ by owning model

### LINKED ISSUE #127: Push model-private kernels down to their model crates, group FFI bindings by domain
## Why

`pegainfer-kernels` currently owns every kernel for every model in one crate. Two related problems have come out of that:

- **`src/ffi.rs` is 1376 lines of `unsafe extern "C"` declarations**, all in a single file. The structure doesn't follow `csrc/` — attention, sampling, norm, fused proj, MoE, FP4/FP8, NCCL extern declarations are all interleaved. Adding a new kernel means finding a spot to insert a block in a giant file; PRs touching kernels regularly hit merge conflicts here.
- **Model-private kernels live in a shared crate that doesn't need them.** Looking at `csrc/`, the boundary between truly shared kernels and model-specific ones is actually quite clean — but the file layout doesn't reflect it, so when a model is added, removed, or refactored, the kernel changes are scattered between `pegainfer-kernels` and the model crate.

Today's actual boundary (just for context — implementer can verify and adjust):

| Shared | Qwen3.5-private | DeepSeek V4-private |
| --- | --- | --- |
| `argmax`, `elementwise`, `fused_proj`, `linear`, `norm`, `paged_attention`, `pos_enc`, `prefill_attention`, `flashinfer_*` | `conv1d`, `gated_delta_rule`, `prefill_attention_hd256`, Triton AOT scripts | entire `csrc/deepseek_v4/` subtree (TileLang + CuTeDSL artifacts) |
| | | |

Qwen3-4B has no truly private kernels — it consumes shared ones.

## What we want

Two changes, ideally in the same PR since they touch overlapping files:

1. **FFI bindings grouped to follow the kernel domain they expose.** A reader looking for the extern for a given `.cu` file should land in a predictable place. Whether that's per-domain Rust files, an auto-generated layout, or something else is up to the implementer.
2. **Model-private kernels (sources, FFI, Rust ops wrappers, build steps) live with the model crate that owns them.** A model's kernel surface — including its build-time generation (Triton for Qwen3.5, TileLang/CuTeDSL for DSV4) — should be removable along with the model crate, not stranded in shared infrastructure.

Implementer can choose: keep `pegainfer-kernels` as the shared kernel home (and possibly rename it to reflect that), or restructure further. The constraint is that the resulting structure should make it obvious where a given kernel lives based on which model it serves.

## Out of scope

- The internal organization of `build.rs` (1306 lines doing 5 things — separate concern).
- Any change to the kernel implementations themselves. This is purely a structural move.
- Cross-toolchain reorganization (CUDA vs. Triton vs. TileLang vs. CuTeDSL as separate sub-crates) — that's a longer-term direction; this issue is the shorter-term untangling.

## Caution

DSV4's kernel surface is the largest and feature-gated. Moving it should keep `--features deepseek-v4` working end-to-end, and the existing E2E test (`Qwen3-8B.json`-style hash regression) should still pass. Same for Qwen3.5's Triton AOT pipeline.

### PR DESCRIPTION
**Pull request by scatyf3**
\## Description

Fixes #127 

### Design Choice

**What this PR is**: **within a single crate, do the owner-grouping** at the filesystem level so the boundary decisions (shared vs private, feature-gate placement, file ownership) are settled and visible.

The Long Term Goal is: moving `pegainfer-kernels/csrc/<model_name>/` and the matching FFI into the `pegainfer-<model_name>` crate, leaving `pegainfer-kernels` with only `shared/`. But our build toolchain runs **CUDA (nvcc) / Triton AOT / TileLang / CuTeDSL** — four codegen pipelines — all wired into `pegainfer-kernels/build.rs`. Pushing kernels into per-model crates today would mean duplicating that entire toolchain across every crate. The prerequisite is extracting a shared build-util crate (e.g. `pegainfer-kernels-build`) that exposes nvcc / Triton / TileLang / CuTeDSL invocations as reusable `[build-dependencies]`. That's a separate piece of work and is out of scope for this PR.

### FFI (`pegainfer-kernels/src/ffi`)

`ffi.rs` becomes a 14-line module shim; the body splits into 4 owner-scoped submodules:

```
pegainfer-kernels/src/ffi.rs        # mod declarations + pub use re-exports
pegainfer-kernels/src/ffi/
├── shared.rs       (42 fns)   # cross-model: rms_norm / embedding / gemm / sampling / paged attention
├── qwen35.rs       (14 fns)   # Qwen3.5 hybrid: conv1d / gated delta / HD256 prefill
├── deepseek.rs     (61 fns)   # DSv4: fp8/fp4 quant / MLA / MoE / compressor / indexer / hc
└── kimi.rs         (29 fns)   # Kimi-K2: MLA / Marlin INT4 / router / experts
                                # 146 FFI symbols total
```

**Public API is unchanged**: `pegainfer_kernels::ffi::<symbol>` still resolves the same way thanks to `pub use shared::*; pub use qwen35::*; ...`. No call site touched.

### csrc

```
pegainfer-kernels/csrc
├── common.cuh                  # shared header, stays at root
├── shared/      (10 .cu)
├── qwen35/      (3  .cu)
├── deepseek_v4/ (6  .cu + deepseek_common.cuh)
└── kimi_k2/     (5  .cu + vendored vllm_marlin/ headers)
```

To adapt the refactor, include path: csrc/ root added to nvcc -I so subdir sources resolve #include "common.cuh" without relative paths.

### CFG Gate Inconsistency 

| Model   | Cargo.toml feature | `csrc/X/` build.rs gating                     | FFI declaration gating          |
| ------- | ------------------ | --------------------------------------------- | ------------------------------- |
| Kimi-K2 | `kimi-k2 = []`     | ✅ build.rs skips `csrc/kimi_k2/` when off     | ✅ `#[cfg(feature = "kimi-k2")]` |
| DSv4    | `deepseek-v4 = []` | ✅ build.rs skips `csrc/deepseek_v4/` when off | ❌ none                          |

The new `ffi.rs` preserves main's existing pattern — Kimi-K2 is fully feature-gated, DSv4 isn't:

```rs
// FFI surface for CUDA/cuBLAS/FlashInfer kernels, split by owning model.
// Public paths are unchanged: `pegainfer_kernels::ffi::<symbol>` resolves via the re-exports below.

// Half type (16-bit float) - same layout as CUDA half. Shared ABI type used by all submodules.
pub type Half = u16;

mod shared;
mod qwen35;
mod deepseek;
#[cfg(feature = "kimi-k2")]
mod kimi;
pub use shared::*;
pub use qwen35::*;
pub use deepseek::*;
#[cfg(feature = "kimi-k2")]
pub use kimi::*;
```

Practical impact: building the workspace without --features deepseek-v4 currently fails to link
(undefined symbol: deepseek_fp8_w1_w3_with_workspace_cuda, etc.) because the .cu files are gated
but the FFI declarations and pegainfer-deepseek-v4 lib are not. Aligning DSv4 with Kimi-K2's pattern
requires gating modules inside pegainfer-deepseek-v4 as well — a cross-crate change kept out of this
PR's scope.

This might be an Good first issue for a newcomer? Should I write this issue and labeled it as good first issue?

## Type of Change

Refactor

## Checklist

-  My code follows the style guidelines of this project (see `docs/areas/coding-style.md`).
-  I have performed a self-review of my own code.
-  I have formatted my commits according to Commitizen conventions.
-  I have run the local test suite and all tests pass (see `CLAUDE.md`).

### CODE REVIEW

**Review by xiaguan — APPROVED** [review]
LGTM

**Comment by xiaguan** [comment]
Most of it LGTM. Please update the PR title to follow the standard format and rebase onto main.

**Review by xiaguan — APPROVED** [review]
LGTM, thanks @scatyf3! Verified beyond the earlier approval (which predated the rebase): byte-level audit of the FFI surface across the four split files — 154 decls on both sides, zero signature drift, zero drops/duplicates; the post-rebase `53524a4` gemm `-> i32` restore matches the `.cu` definitions exactly, and the module-level kimi cfg-gate is cleaner than main's per-block gates. Ran the reorganized tree on a local GPU: full workspace lib tests plus the qwen3-4b suite (hf_golden_gate, prefix_cache, scheduler_robustness) all pass — the moved `shared/` kernels produce HF-parity logits end-to-end. One note for the record: the lone per-fn `deepseek-v4` gate on `deepseek_cutedsl_indexer_scores_exact_bf16_cuda` is now uniformly ungated with its siblings — decl-only, no link impact, and consistent with the documented DSv4 FFI stance.
